### ISSUE TITLE
Introduce new STOP_AT parameter (in addition to START_FROM)

### ISSUE THREAD

**Issue by hiaik** [gitlabform/gitlabform#877]
**What:**
We would like to have `STOP_AT` parameter to specify which project should be processed as last (all the rest are skipped).

**Current behaviour:**
Only start-from is supported which is unclear on how to use (what was the purpose of that feature)
```
  -sf START_FROM, --start-from START_FROM
                        start processing projects from the given one (as numbered by "x/y Processing group/project" messages) (default: 1)
```

**Why:**
We need to run gitlabform project for ~600 projects hourly. On gitlabform 4.0.1 version time jumped up (from 40 minutes to 1.5 hour) in pipeline as there was a logic change in how API are executed. It will be more and more in the future.

Idea is to be able to run the same `config.yml` in parallel (in batches):
* gitlabform -c config.yml ${PROJECT_PATH} --start-from 1 --stop-at 100
* gitlabform -c config.yml ${PROJECT_PATH} --start-from 101 --stop-at 200
...
* gitlabform -c config.yml ${PROJECT_PATH} --start-from N --stop-at M

That way we can process all projects in batches via single config file.

Are you ok if we contribute to that feature?

**Comment by amimas** [comment]
Hi @hiaik - Contribution is always welcome. I do have some thoughts before that is continued though. Would love to get feedbacks.

First, let's address the issue with a jump in execution time. This came from the fix that was added via PR #857. There's a 2 second wait time added in branch protection config for each branch when the project also contains member config. Reason for that is due to a GitLab Issue, which is explained in that PR. So in your case, 2 seconds for 600 projects is about 1200 seconds (20 minutes) waiting time. If there were multiple branches being configured, it will take longer. So, that explains the execution time increase for you from 40 minutes to 90 minutes. I just opened another PR #878 that should reduce the time a little bit, but I imagine you'll still see total execution time at around 60 minutes.

I think the GitLab issue mentioned/found in that PR should be addressed. If someone can work with GitLab support and lead the effort on that, it would be best solution in general. That way we don't need to add that wait time unnecessarily. There was a related issue in GitLab, but it was simply closed. Probably didn't come into the radar of appropriate folks.

https://gitlab.com/gitlab-org/gitlab/-/issues/383666#note_1338023609

As for your suggestion about `STOP_AT` parameter, it makes sense since there is one related parameter called `START_FROM`. I'm also not sure what was the purpose of "start-from". But assuming both of these are available, the way you're thinking of using this feature, I wonder if you will run into API rate limits due to many API calls in parallel. I have run gitlabform in parallel in my usage before, but the number of projects was smaller than yours. If it's GitLab Saas, you may not have any control over the API rate limit.

Regardless, I was actually wondering whether the `start_from` should be dropped. Keep gitlabform a simple CLI focusing on applying configurations. Orchestration could probably be done using separate tools/scripts by the users.

What do you think @gdubicki @TimKnight-DWP ?

**Comment by hiaik** [comment]
Thanks @amimas for all the input.

> I just opened another PR https://github.com/gitlabform/gitlabform/pull/878 that should reduce the time a little bit, but I imagine you'll still see total execution time at around 60 minutes.

Thanks, I will monitor that and will update to the latest version once it is merged.
Any quick win is always good.
I understand on why sleep(2) was introduced. From what I can see gitlab has lots of background jobs which might take time.


> I think the GitLab issue mentioned/found in that PR should be addressed. If someone can work with GitLab support and lead the effort on that, it would be best solution in general. That way we don't need to add that wait time unnecessarily. There was a related issue in GitLab, but it was simply closed. Probably didn't come into the radar of appropriate folks.
https://gitlab.com/gitlab-org/gitlab/-/issues/383666#note_1338023609

I am happy to start convo with gitlab support on that as we have ultimate licence and they can hear us better.
I need to replicate issue on my side first. Could you please help with that?

At this stage what I did is:

1. I had a configuration for
```
  "project_path":
    members:
      groups:
        group_path:
          group_access: 30
    merge_requests_approval_rules:
      develop:
        approvals_required: 1
        name: "develop branch"
        groups:
          - group_path
        protected_branches:
          - develop
```

2. I ran it with debug
```
docker run -it -e GITLAB_TOKEN=... -v $PWD:/app -w /app ghcr.io/gitlabform/gitlabform:4.0.1 sh -c 'gitlabform -c projects-config.yml -d "project_path"'
```

to see which calls are executing


3. I tried to run my shell script as below:
```
echo "DELETE share \n"
curl --location --request DELETE 'https://gitlab.com/api/v4/projects/project_path/share/90651443' \
--header 'PRIVATE-TOKEN: ...' \
--header 'Cookie: ...'

echo "POST share \n"
curl --location 'https://gitlab.com/api/v4/projects/project_path/share' \
--header 'PRIVATE-TOKEN: ...' \
--header 'Content-Type: application/json' \
--header 'Cookie: ...' \
--data '{
    "expires_at": null,
    "group_access": 30,
    "group_id": 90651443
}'

echo "POST approval_rules \n"
curl --location 'https://gitlab.com/api/v4/projects/60392570/approval_rules' \
--header 'PRIVATE-TOKEN: ...' \
--header 'Content-Type: application/json' \
--header 'Cookie: ...' \
--data '{
    "approvals_required": 1,
    "group_ids": [
        90651443
    ],
    "name": "develop branch",
    "protected_branch_ids": [
        130216665
    ]
}'

```

However, it all works well -> no error. 

Could you please help me to replicate that error and I am happy to report it to GitLab with context for existing issue:
https://gitlab.com/gitlab-org/gitlab/-/issues/383666#note_1338023609




> I wonder if you will run into API rate limits due to many API calls in parallel. I have run gitlabform in parallel in my usage before, but the number of projects was smaller than yours. If it's GitLab Saas, you may not have any control over the API rate limit.

limits look pretty good https://docs.gitlab.com/ee/user/gitlab_com/index.html#gitlabcom-specific-rate-limits and we are running some of the projects right now in parallel (we split them by project path passed to `gitlabform`). It is just an issue that one project path has ~400 projects in it.



>Regardless, I was actually wondering whether the start_from should be dropped. Keep gitlabform a simple CLI focusing on applying configurations. Orchestration could probably be done using separate tools/scripts by the users.

Not sure tbh how we can orchestrate effective configuration created by `gitlabform` (specifically for my request: when gitlabform generates, for example, 400 projects config -> numbered it 1..400 -> now we just have to split it in batches and process). It sounds inefficient to me to use another tool for that.

**Comment by amimas** [comment]
> Not sure tbh how we can orchestrate effective configuration created by gitlabform (specifically for my request: when gitlabform generates, for example, 400 projects config -> numbered it 1..400 -> now we just have to split it in batches and process). It sounds inefficient to me to use another tool for that.

I agree. I was also thinking about it earlier and couldn't think of an alternative. In my past usage, I basically had separate config file for different groups. Then each of those config files were used in different jobs in parallel. Anyways, let's add the `stop_at` feature to the CLI. Could you please help contribute it?

As for the other issue, I'll see if I can reproduce it.

**Comment by hiaik** [comment]
> In my past usage, I basically had separate config file for different groups. Then each of those config files were used in different jobs in parallel. Anyways, let's add the stop_at feature to the CLI. 

We have single config file where the majority of configuration is the same (re-use via yaml anchors). And once you have around 400 projects in one gitlab path/group, it is inevitable to face performance issue (split files with the same configuration will complicate maintenance as against DRY principle, etc.)


> Could you please help contribute it?

With a great pleasure. I have planned it to backlog. Will update once have time to work on it.

**Comment by TimKnight-DWP** [comment]
@hiaik @amimas - If you do re-raise something with GitLab (it is on their end for sure, they do a lot of async processing now), please let me know and I can bother our Ultimate Reps to escalate. I think we see similar API issues our end that are transitive, just not had time to investigate

**Comment by amimas** [comment]
@hiaik - after a bit more digging trying to reproduce the issue, I realized that it only happens when new users are being created and then they are being referenced in various settings (i.e. branch protection, environment protection, etc.). Since gitlabform doesn't actually "create gitlab users" (as of now), this is only an issue for the automated tests because the test setup creates new user and then runs gitlabform. All of these runs very quickly and in that scenario trying to configure settings such as those mentioned above, runs into issues.

I just opened PR #885 to remove the delay that was added in the runtime previously in v4.0.1 trying to address #765. The delay is only needed in gitlabform's acceptance tests. Hopefully we can release it soon. This should go back to the execution time prior to v4.0.1.

**Comment by hiaik** [comment]
thanks @amimas 4.0.3 returned execution time to how it was before on 3.16.5 etc.
All good.


> I realized that it only happens when new users are being created and then they are being referenced in various settings (i.e. branch protection, environment protection, etc.). Since gitlabform doesn't actually "create gitlab users" (as of now), this is only an issue for the automated tests because the test setup creates new user and then runs gitlabform.

ok, so nothing to follow up with gitlab at this stage.



This ticket I would leave to continue working on STOP_AT feature
